`element types must match for conversion to set` error when using set of object with optional map with optional attribute.

### Terraform Version

```shell
Terraform v1.10.3
on linux_amd64
```


### Terraform Configuration Files

```
variable "tenants" {
  type = set(
    object(
      {
        optional_map = optional(map( object({
            asdf = optional(string, "aaa")
        })), {})
      }
    )
  )
}
```

### Debug Output
```
2024-12-27T07:39:25.066-0500 [INFO]  Terraform version: 1.10.3
2024-12-27T07:39:25.067-0500 [DEBUG] using github.com/hashicorp/go-tfe v1.70.0
2024-12-27T07:39:25.067-0500 [DEBUG] using github.com/hashicorp/hcl/v2 v2.23.0
2024-12-27T07:39:25.067-0500 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2024-12-27T07:39:25.067-0500 [DEBUG] using github.com/zclconf/go-cty v1.15.1-0.20241111215639-63279be090d7
2024-12-27T07:39:25.067-0500 [INFO]  Go runtime version: go1.23.3
2024-12-27T07:39:25.067-0500 [INFO]  CLI args: []string{"terraform", "plan"}
2024-12-27T07:39:25.067-0500 [TRACE] Stdout is a terminal of width 383
2024-12-27T07:39:25.067-0500 [TRACE] Stderr is a terminal of width 383
2024-12-27T07:39:25.067-0500 [TRACE] Stdin is a terminal
2024-12-27T07:39:25.067-0500 [DEBUG] Attempting to open CLI config file: /home/bellis/.terraformrc
2024-12-27T07:39:25.067-0500 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2024-12-27T07:39:25.067-0500 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2024-12-27T07:39:25.067-0500 [DEBUG] ignoring non-existing provider search directory /home/bellis/.terraform.d/plugins
2024-12-27T07:39:25.067-0500 [DEBUG] ignoring non-existing provider search directory /home/bellis/.local/share/terraform/plugins
2024-12-27T07:39:25.067-0500 [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
2024-12-27T07:39:25.067-0500 [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
2024-12-27T07:39:25.067-0500 [DEBUG] ignoring non-existing provider search directory /var/lib/snapd/desktop/terraform/plugins
2024-12-27T07:39:25.067-0500 [INFO]  CLI command args: []string{"plan"}
2024-12-27T07:39:25.067-0500 [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2024-12-27T07:39:25.067-0500 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2024-12-27T07:39:25.067-0500 [TRACE] Meta.Backend: using default local state only (no backend configuration, and no existing initialized backend)
2024-12-27T07:39:25.067-0500 [TRACE] Meta.Backend: instantiated backend of type <nil>
2024-12-27T07:39:25.068-0500 [DEBUG] checking for provisioner in "."
2024-12-27T07:39:25.069-0500 [DEBUG] checking for provisioner in "/usr/bin"
2024-12-27T07:39:25.069-0500 [TRACE] Meta.Backend: backend <nil> does not support operations, so wrapping it in a local backend
2024-12-27T07:39:25.069-0500 [INFO]  backend/local: starting Plan operation
2024-12-27T07:39:25.069-0500 [TRACE] backend/local: requesting state manager for workspace "default"
2024-12-27T07:39:25.069-0500 [TRACE] backend/local: state manager for workspace "default" will:
 - read initial snapshot from terraform.tfstate
 - write new snapshots to terraform.tfstate
 - create any backup at terraform.tfstate.backup
2024-12-27T07:39:25.069-0500 [TRACE] backend/local: requesting state lock for workspace "default"
2024-12-27T07:39:25.069-0500 [TRACE] statemgr.Filesystem: preparing to manage state snapshots at terraform.tfstate
2024-12-27T07:39:25.069-0500 [TRACE] statemgr.Filesystem: no previously-stored snapshot exists
2024-12-27T07:39:25.069-0500 [TRACE] statemgr.Filesystem: locking terraform.tfstate using fcntl flock
2024-12-27T07:39:25.070-0500 [TRACE] statemgr.Filesystem: writing lock metadata to .terraform.tfstate.lock.info
2024-12-27T07:39:25.070-0500 [TRACE] backend/local: reading remote state for workspace "default"
2024-12-27T07:39:25.070-0500 [TRACE] statemgr.Filesystem: reading latest snapshot from terraform.tfstate
2024-12-27T07:39:25.070-0500 [TRACE] statemgr.Filesystem: snapshot file has nil snapshot, but that's okay
2024-12-27T07:39:25.070-0500 [TRACE] statemgr.Filesystem: read nil snapshot
2024-12-27T07:39:25.070-0500 [TRACE] backend/local: populating backendrun.LocalRun for current working directory
2024-12-27T07:39:25.070-0500 [TRACE] terraform.NewContext: starting
2024-12-27T07:39:25.070-0500 [TRACE] terraform.NewContext: complete
2024-12-27T07:39:25.070-0500 [TRACE] backend/local: requesting interactive input, if necessary
2024-12-27T07:39:25.070-0500 [TRACE] Context.Input: Prompting for provider arguments
2024-12-27T07:39:25.070-0500 [TRACE] backend/local: running validation operation
2024-12-27T07:39:25.070-0500 [DEBUG] Building and walking validate graph
2024-12-27T07:39:25.070-0500 [TRACE] building graph for walkValidate
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.ConfigTransformer
2024-12-27T07:39:25.070-0500 [TRACE] ConfigTransformer: Starting for path:
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.ConfigTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
  var.tenants - *terraform.NodeRootVariable
  ------
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.variableValidationTransformer
2024-12-27T07:39:25.070-0500 [TRACE] variableValidationTransformer: adding validation nodes for any existing variable evaluation nodes
2024-12-27T07:39:25.070-0500 [TRACE] variableValidationTransformer: var.tenants has no validation rules
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.variableValidationTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.LocalTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.OutputTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.OutputTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.checkTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.checkTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.StateTransformer
2024-12-27T07:39:25.070-0500 [TRACE] StateTransformer: pointless no-op call, creating no nodes at all
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.externalProviderTransformer
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.externalProviderTransformer with new graph:
  var.tenants - *terraform.NodeRootVariable
  ------
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2024-12-27T07:39:25.070-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.graphTransformerMulti (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.ModuleExpansionTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.ModuleExpansionTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.ExternalReferenceTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.ExternalReferenceTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2024-12-27T07:39:25.070-0500 [DEBUG] ReferenceTransformer: "var.tenants" references: []
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.ReferenceTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.AttachDependenciesTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.AttachDependenciesTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.attachDataResourceDependsOnTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.attachDataResourceDependsOnTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.DestroyEdgeTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.DestroyEdgeTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.pruneUnusedNodesTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.pruneUnusedNodesTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.TargetsTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.ForcedCBDTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.ForcedCBDTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.ephemeralResourceCloseTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.ephemeralResourceCloseTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.CloseProviderTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.CloseProviderTransformer (no changes)
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.CloseRootModuleTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.CloseRootModuleTransformer with new graph:
  root - *terraform.nodeCloseModule
    var.tenants - *terraform.NodeRootVariable
  var.tenants - *terraform.NodeRootVariable
  ------
2024-12-27T07:39:25.070-0500 [TRACE] Executing graph transform *terraform.TransitiveReductionTransformer
2024-12-27T07:39:25.070-0500 [TRACE] Completed graph transform *terraform.TransitiveReductionTransformer (no changes)
2024-12-27T07:39:25.070-0500 [DEBUG] Starting graph walk: walkValidate
2024-12-27T07:39:25.070-0500 [TRACE] vertex "var.tenants": starting visit (*terraform.NodeRootVariable)
2024-12-27T07:39:25.070-0500 [TRACE] vertex "var.tenants": belongs to
2024-12-27T07:39:25.070-0500 [TRACE] NodeRootVariable: evaluating var.tenants
2024-12-27T07:39:25.070-0500 [TRACE] prepareFinalInputVariableValue: preparing var.tenants
2024-12-27T07:39:25.070-0500 [TRACE] vertex "var.tenants": visit complete
2024-12-27T07:39:25.070-0500 [TRACE] vertex "root": starting visit (*terraform.nodeCloseModule)
2024-12-27T07:39:25.070-0500 [TRACE] vertex "root": does not belong to any module instance
2024-12-27T07:39:25.070-0500 [TRACE] vertex "root": visit complete
2024-12-27T07:39:25.070-0500 [INFO]  backend/local: plan calling Plan
2024-12-27T07:39:25.071-0500 [DEBUG] Building and walking plan graph for NormalMode
2024-12-27T07:39:25.071-0500 [TRACE] building graph for walkPlan
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.ConfigTransformer
2024-12-27T07:39:25.071-0500 [TRACE] ConfigTransformer: Starting for path:
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.ConfigTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
  var.tenants - *terraform.NodeRootVariable
  ------
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.variableValidationTransformer
2024-12-27T07:39:25.071-0500 [TRACE] variableValidationTransformer: adding validation nodes for any existing variable evaluation nodes
2024-12-27T07:39:25.071-0500 [TRACE] variableValidationTransformer: var.tenants has no validation rules
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.variableValidationTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.LocalTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.OutputTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.OutputTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.checkTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.checkTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.StateTransformer
2024-12-27T07:39:25.071-0500 [TRACE] StateTransformer: creating nodes for deposed instance objects only
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.externalProviderTransformer
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.externalProviderTransformer with new graph:
  var.tenants - *terraform.NodeRootVariable
  ------
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2024-12-27T07:39:25.071-0500 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.graphTransformerMulti (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.ModuleExpansionTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.ModuleExpansionTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.ExternalReferenceTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.ExternalReferenceTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2024-12-27T07:39:25.071-0500 [DEBUG] ReferenceTransformer: "var.tenants" references: []
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.ReferenceTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.AttachDependenciesTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.AttachDependenciesTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.attachDataResourceDependsOnTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.attachDataResourceDependsOnTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.DestroyEdgeTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.DestroyEdgeTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.pruneUnusedNodesTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.pruneUnusedNodesTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.TargetsTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.ForcedCBDTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.ForcedCBDTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.ephemeralResourceCloseTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.ephemeralResourceCloseTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.CloseProviderTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.CloseProviderTransformer (no changes)
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.CloseRootModuleTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.CloseRootModuleTransformer with new graph:
  root - *terraform.nodeCloseModule
    var.tenants - *terraform.NodeRootVariable
  var.tenants - *terraform.NodeRootVariable
  ------
2024-12-27T07:39:25.071-0500 [TRACE] Executing graph transform *terraform.TransitiveReductionTransformer
2024-12-27T07:39:25.071-0500 [TRACE] Completed graph transform *terraform.TransitiveReductionTransformer (no changes)
2024-12-27T07:39:25.071-0500 [DEBUG] Starting graph walk: walkPlan
2024-12-27T07:39:25.071-0500 [TRACE] vertex "var.tenants": starting visit (*terraform.NodeRootVariable)
2024-12-27T07:39:25.071-0500 [TRACE] vertex "var.tenants": belongs to
2024-12-27T07:39:25.071-0500 [TRACE] NodeRootVariable: evaluating var.tenants
2024-12-27T07:39:25.071-0500 [TRACE] prepareFinalInputVariableValue: preparing var.tenants
2024-12-27T07:39:25.071-0500 [ERROR] prepareFinalInputVariableValue: var.tenants has unsuitable type
  got:  {{{} [{{{} map[optional_map:{{{} map[] map[]}}] map[]}} {{{} map[optional_map:{{{} {{{} map[asdf:{{{} %!s(cty.primitiveTypeKind=83)}}] map[]}}}}] map[]}}]}}
  want: {{{} {{{} map[optional_map:{{{} {{{} map[asdf:{{{} %!s(cty.primitiveTypeKind=83)}}] map[asdf:{}]}}}}] map[optional_map:{}]}}}}
2024-12-27T07:39:25.071-0500 [ERROR] vertex "var.tenants" error: Invalid value for input variable
2024-12-27T07:39:25.071-0500 [TRACE] vertex "var.tenants": visit complete, with errors
2024-12-27T07:39:25.071-0500 [TRACE] dag/walk: upstream of "root" errored, so skipping
2024-12-27T07:39:25.071-0500 [WARN]  Planning encountered errors, so plan is not applyable
2024-12-27T07:39:25.071-0500 [INFO]  backend/local: plan operation complete
Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Invalid value for input variable
│ 
│   on terraform.tfvars line 1:
│    1: tenants = [
│    2:     {
│    3:         optional_map = {}
│    4:     },
│    5:     {
│    6:     }
│    7: ]
│ 
│ The given value is not suitable for var.tenants declared at main.tf:2,1-19: element types must all match for conversion to set.
╵
2024-12-27T07:39:25.071-0500 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
2024-12-27T07:39:25.071-0500 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
```

### Expected Behavior

No issues when generating plan

### Actual Behavior

Planning failed. Terraform encountered an error while generating this plan.

```
╷
│ Error: Invalid value for input variable
│ 
│   on terraform.tfvars line 1:
│    1: tenants = [
│    2:     {
│    3:         optional_map = {}
│    4:     },
│    5:     {
│    6:     }
│    7: ]
│ 
│ The given value is not suitable for var.tenants declared at main.tf:2,1-19: element types must all match for conversion to set.
```

### Steps to Reproduce

1. `terraform plan`

### Additional Context

Here is the provided terraform.tfvars

```
tenants = [
    {
        optional_map = {}
    },
    {
    }
]
```

### References

Related, but should have been fixed 2 years ago?
https://github.com/hashicorp/terraform/issues/31924


### Generative AI / LLM assisted development?

no

--- a/go.mod
+++ b/go.mod
@@ -61,7 +61,7 @@ require (
 	github.com/spf13/afero v1.9.5
 	github.com/xanzy/ssh-agent v0.3.3
 	github.com/xlab/treeprint v0.0.0-20161029104018-1d6e34225557
-	github.com/zclconf/go-cty v1.15.1
+	github.com/zclconf/go-cty v1.16.0
 	github.com/zclconf/go-cty-debug v0.0.0-20240509010212-0d6042c53940
 	github.com/zclconf/go-cty-yaml v1.1.0
 	go.opentelemetry.io/contrib/exporters/autoexport v0.45.0

--- a/internal/backend/remote-state/azure/go.mod
+++ b/internal/backend/remote-state/azure/go.mod
@@ -50,7 +50,7 @@ require (
 	github.com/mitchellh/reflectwalk v1.0.2 // indirect
 	github.com/spf13/afero v1.9.5 // indirect
 	github.com/stretchr/testify v1.8.4 // indirect
-	github.com/zclconf/go-cty v1.15.1 // indirect
+	github.com/zclconf/go-cty v1.16.0 // indirect
 	golang.org/x/crypto v0.31.0 // indirect
 	golang.org/x/mod v0.21.0 // indirect
 	golang.org/x/net v0.29.0 // indirect

--- a/internal/backend/remote-state/consul/go.mod
+++ b/internal/backend/remote-state/consul/go.mod
@@ -6,7 +6,7 @@ require (
 	github.com/hashicorp/consul/api v1.13.0
 	github.com/hashicorp/consul/sdk v0.8.0
 	github.com/hashicorp/terraform v0.0.0-00010101000000-000000000000
-	github.com/zclconf/go-cty v1.15.1
+	github.com/zclconf/go-cty v1.16.0
 )
 
 require (

--- a/internal/backend/remote-state/cos/go.mod
+++ b/internal/backend/remote-state/cos/go.mod
@@ -32,7 +32,7 @@ require (
 	github.com/mitchellh/reflectwalk v1.0.2 // indirect
 	github.com/mozillazg/go-httpheader v0.3.0 // indirect
 	github.com/spf13/afero v1.9.5 // indirect
-	github.com/zclconf/go-cty v1.15.1 // indirect
+	github.com/zclconf/go-cty v1.16.0 // indirect
 	golang.org/x/mod v0.21.0 // indirect
 	golang.org/x/net v0.29.0 // indirect
 	golang.org/x/sync v0.10.0 // indirect

--- a/internal/backend/remote-state/gcs/go.mod
+++ b/internal/backend/remote-state/gcs/go.mod
@@ -7,7 +7,7 @@ require (
 	cloud.google.com/go/storage v1.30.1
 	github.com/hashicorp/terraform v0.0.0-00010101000000-000000000000
 	github.com/mitchellh/go-homedir v1.1.0
-	github.com/zclconf/go-cty v1.15.1
+	github.com/zclconf/go-cty v1.16.0
 	golang.org/x/oauth2 v0.23.0
 	google.golang.org/api v0.126.0
 	google.golang.org/genproto v0.0.0-20230822172742-b8732ec3820d

--- a/internal/backend/remote-state/kubernetes/go.mod
+++ b/internal/backend/remote-state/kubernetes/go.mod
@@ -5,7 +5,7 @@ go 1.23.3
 require (
 	github.com/hashicorp/terraform v0.0.0-00010101000000-000000000000
 	github.com/mitchellh/go-homedir v1.1.0
-	github.com/zclconf/go-cty v1.15.1
+	github.com/zclconf/go-cty v1.16.0
 	k8s.io/api v0.25.5
 	k8s.io/apimachinery v0.25.5
 	k8s.io/client-go v0.25.5

--- a/internal/backend/remote-state/oss/go.mod
+++ b/internal/backend/remote-state/oss/go.mod
@@ -39,7 +39,7 @@ require (
 	github.com/satori/go.uuid v1.2.0 // indirect
 	github.com/spf13/afero v1.9.5 // indirect
 	github.com/stretchr/testify v1.8.4 // indirect
-	github.com/zclconf/go-cty v1.15.1 // indirect
+	github.com/zclconf/go-cty v1.16.0 // indirect
 	golang.org/x/mod v0.21.0 // indirect
 	golang.org/x/net v0.29.0 // indirect
 	golang.org/x/sync v0.10.0 // indirect

--- a/internal/backend/remote-state/pg/go.mod
+++ b/internal/backend/remote-state/pg/go.mod
@@ -7,7 +7,7 @@ require (
 	github.com/hashicorp/hcl/v2 v2.23.0
 	github.com/hashicorp/terraform v0.0.0-00010101000000-000000000000
 	github.com/lib/pq v1.10.3
-	github.com/zclconf/go-cty v1.15.1
+	github.com/zclconf/go-cty v1.16.0
 )
 
 require (

--- a/internal/backend/remote-state/s3/go.mod
+++ b/internal/backend/remote-state/s3/go.mod
@@ -15,7 +15,7 @@ require (
 	github.com/hashicorp/go-uuid v1.0.3
 	github.com/hashicorp/hcl/v2 v2.23.0
 	github.com/hashicorp/terraform v0.0.0-00010101000000-000000000000
-	github.com/zclconf/go-cty v1.15.1
+	github.com/zclconf/go-cty v1.16.0
 )
 
 require (

--- a/internal/legacy/go.mod
+++ b/internal/legacy/go.mod
@@ -11,7 +11,7 @@ require (
 	github.com/mitchellh/copystructure v1.2.0
 	github.com/mitchellh/mapstructure v1.5.0
 	github.com/mitchellh/reflectwalk v1.0.2
-	github.com/zclconf/go-cty v1.15.1
+	github.com/zclconf/go-cty v1.16.0
 )
 
 require (

